### PR TITLE
Kernel version fixes and null pointer fix

### PR DESCRIPTION
**Pull request by BobBeckett**
Fix building on latest upstream kernel
Fix module unload and remove_all when no fbs have been created.

### CODE REVIEW

File: module/evdi_drv.h [hunk 1]
```diff
@@ -22,7 +22,11 @@
 #include <drm/drm_crtc_helper.h>
 #include <drm/drm_rect.h>
 #include <drm/drm_gem.h>
+#if KERNEL_VERSION(5, 3, 0) >= LINUX_VERSION_CODE
```
Comment (mabrand) [inline module/evdi_drv.h]: Isn't the new situation from 5.4.0? This change breaks build for kernel 5.3.18 at least.

File: module/evdi_connector.c [hunk 2]
```diff
@@ -109,11 +109,21 @@ static void evdi_connector_destroy(struct drm_connector *connector)
 static struct drm_encoder *evdi_best_encoder(struct drm_connector *connector)
 {
 
+#if KERNEL_VERSION(5, 4, 0) >= LINUX_VERSION_CODE
```
Comment (s0be) [inline module/evdi_connector.c]: This needs to be <=

**Comment by displaylink-dkurek** [comment]
Hi,
this would be best to split into two PR's: 
* null pointer fix
* kernel 4.4 compatibility
First looks like something that could go in as it is now. Is this something you have encountered (that the pointer was null) or at this point it's purely theoretical at this point?

About the `#if`s it should be 
> #if KERNEL_VERSION(X, Y, Z) <= LINUX_VERSION_CODE

or

> #if KERNEL_VERSION(X, Y, Z) > LINUX_VERSION_CODE

We probably could add a check for this to CI... :)

**Comment by abucodonosor** [comment]
> I'd suggest the following change since it doesn't build on 5.4.x
> 5.5 have a different syntax here since commit 62afb4ad425af
> 
> ```
>     /* return the first found */
> 
>     #if KERNEL_VERSION(5, 5, 0) <= LINUX_VERSION_CODE
> 
>       drm_connector_for_each_possible_encoder(connector, encoder)
> 
>    #else
> 
>      int i;
> 
>      drm_connector_for_each_possible_encoder(connector, encoder, i)
> 
>    #endif
> ```
> 
> `

Well the >= has to be <= and then it should build, your suggestion seems wrong.

On 5.4 the old code should remain the same is just a matter of right defines which btw is a mess in the code. That is the only module I know uses the KERNEL_VERSION & LINUX_VERSION_CODE swapped for checks, therefore the logic on that is backward.

**Comment by displaylink-dkurek** [comment]
I agree that the logic is the wrong way. We have adopted this conversion for
the sake of `checkpath.pl` and it's
```
WARN("CONSTANT_COMPARISON",
     "Comparisons should place the constant on the right side of the test\n" . $herecurr) &&
$fix)
```
check. I know those `#if`s wouldn't get to kernel anyway (and we are disabling 
`LINUX_VERSION_CODE` already) but disabling `CONTANT_COMPARISON` you disable 
it for all comparisons, not only this.

**Comment by abucodonosor** [comment]
@displaylink-dkurek 

Sure, I figured that :-).
One needs to get used to the logic, I was confused the first time I looked into the code too.

**Comment by displaylink-dkurek** [comment]
@BobBeckett,
You are absolutely right about the null pointer. It is not theoretical, it would happen..
Can you please rebase on top of the devel branch, or just create new PR with the
null pointer fix? Thanks.

**Comment by displaylink-dkurek** [comment]
I've introduced the fix for null pointer as commit a7c86de6b56917b2c7c4f4f5773c6fd8da695b8d.
Thanks for this!.

I think now all the changes from this PR has landed. 
Closing

**Pull request closed without merging**